buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java

```diff
@@ -33,73 +33,141 @@
-
+    private void creteUseClusterTaskExtension(Project project) {
         // register an extension for all current and future tasks, so that any task can declare that it wants to use a
         // specific cluster.
         project.getTasks().all((Task task) ->
             task.getExtensions().findByType(ExtraPropertiesExtension.class)
-            .set(
-                "useCluster",
-                new Closure<Void>(this, this) {
-                    public void doCall(ElasticsearchNode conf) {
-                        taskToCluster.computeIfAbsent(task, k -> new ArrayList<>()).add(conf);
-                    }
-                })
+                .set(
+                    "useCluster",
+                    new Closure<Void>(this, task) {
+                        public void doCall(ElasticsearchNode node) {
+                            taskToCluster.computeIfAbsent(task, k -> new ArrayList<>()).add(node);
+                        }
+                    })
         );
+    }
 
+    private void configureClaimClustersHook(Project project) {
         project.getGradle().getTaskGraph().whenReady(taskExecutionGraph ->
             taskExecutionGraph.getAllTasks()
                 .forEach(task ->
                     taskToCluster.getOrDefault(task, Collections.emptyList()).forEach(ElasticsearchNode::claim)
                 )
         );
+    }
+
+    private void configureStartClustersHook(Project project) {
         project.getGradle().addListener(
             new TaskActionListener() {
                 @Override
                 public void beforeActions(Task task) {
                     // we only start the cluster before the actions, so we'll not start it if the task is up-to-date
-                    taskToCluster.getOrDefault(task, new ArrayList<>()).forEach(ElasticsearchNode::start);
+                    taskToCluster.getOrDefault(task, Collections.emptyList()).forEach(ElasticsearchNode::start);
                 }
                 @Override
                 public void afterActions(Task task) {}
             }
         );
+    }
+
+    private void configureStopClustersHook(Project project) {
         project.getGradle().addListener(
             new TaskExecutionListener() {
                 @Override
                 public void afterExecute(Task task, TaskState state) {
-                    // always un-claim the cluster, even if _this_ task is up-to-date, as others might not have been and caused the
-                    // cluster to start.
-                    taskToCluster.getOrDefault(task, new ArrayList<>()).forEach(ElasticsearchNode::unClaimAndStop);
+                    // always unclaim the cluster, even if _this_ task is up-to-date, as others might not have been
+                    // and caused the cluster to start.
+                    if (state.getFailure() != null) {
+                        // If the task fails, and other tasks use this cluster, the other task will likely never be
```

Comment:
This could only happen if a cluster is shared across projects, but that shouldn't happen (and should be forbidden)?

Reply:
The first part can happen when there are multiple tasks using a cluster within the same project 
and one of them fails mid-way. The rest in the chain will not run because Gradle fails fast, but that also means that claims will never reach zero, so we need to stop the cluster and show logs right away. 

Moving the  claims and started logic outside of `ElasticserachNode`  will make this more straight forward as you suggested.

The second part of my comment can indeed only happen with multi project builds.
When two tasks run concurrently and one of them fails, the cluster will be force stopped without regard for the other.

We don't have a _technical_ reason  to  forbid using test clusters across projects 
and the syntax to do it is not that straight forward. 
I would not spend time on implementing logic to forbid it now, we can wait for evidence of abuse.